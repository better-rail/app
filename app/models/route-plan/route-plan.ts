import { Instance, SnapshotOut, types } from "mobx-state-tree"
import { omit } from "ramda"
import { translate } from "../../i18n"

const StationSchema = {
  id: types.string,
  name: types.string,
}

export const RoutePlanModel = types
  .model("RoutePlan")
  .props({
    origin: types.maybe(types.model(StationSchema)),
    destination: types.maybe(types.model(StationSchema)),
    date: types.optional(types.Date, () => new Date()),
    dateType: types.optional(types.enumeration(["departure", "arrival"]), "departure"),
  })
  .views((self) => {
    return {
      get dateTypeDisplayName() {
        if (self.dateType === "departure") return translate("plan.departureTime")
        return translate("plan.arrivalTime")
      },
    }
  })

  .actions((self) => ({
    setOrigin(station) {
      self.origin = station
    },
    setDestination(station) {
      self.destination = station
    },
    setDate(date) {
      self.date = date
    },
    setDateType(type) {
      self.dateType = type
    },
    switchDirection() {
      // Handle cases where the origin/destination are undefined (for example, on initial app launch)
      const origin = self.origin ? Object.assign({}, self.origin) : undefined
      const destination = self.destination ? Object.assign({}, self.destination) : undefined

      this.setOrigin(destination)
      this.setDestination(origin)
    },
    switchDateType() {
      if (self.dateType === "departure") {
        this.setDateType("arrival")
      } else {
        this.setDateType("departure")
      }
    },
  }))
  .postProcessSnapshot(omit(["date"]))

type RoutePlanType = Instance<typeof RoutePlanModel>
export interface RoutePlan extends RoutePlanType {}
type RoutePlanSnapshotType = SnapshotOut<typeof RoutePlanModel>
export interface RoutePlanSnapshot extends RoutePlanSnapshotType {}
export const createRoutePlanDefaultModel = () => types.optional(RoutePlanModel, {})
