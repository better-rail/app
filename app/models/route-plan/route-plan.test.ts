import { getSnapshot } from "mobx-state-tree"
import { RoutePlanModel } from "./route-plan"

test("can be created", () => {
  const origin = {
    name: "ירושלים - יצחק נבון",
    id: "680",
  }

  const destination = {
    name: "הרצליה",
    id: "3500",
  }

  const instance = RoutePlanModel.create({ origin, destination })

  expect(getSnapshot(instance)).toMatchSnapshot()
})
