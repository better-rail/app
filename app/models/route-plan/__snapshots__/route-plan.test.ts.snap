// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be created 1`] = `
{
  "destination": {
    "id": "3500",
    "name": "הרצליה",
  },
  "origin": {
    "id": "680",
    "name": "ירושלים - יצחק נבון",
  },
}
`;
